apps/bluetooth/js/modules/transfer_manager.js

```diff
@@ -0,0 +1,372 @@
+/* -*- Mode: js; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- /
+/* vim: set shiftwidth=2 tabstop=2 autoindent cindent expandtab: */
+
+define(function(require) {
+  'use strict';
+
+  var AdapterManager = require('modules/bluetooth/bluetooth_adapter_manager');
+  var BtContext = require('modules/bluetooth/bluetooth_context');
+  var CannotTransferDialog = require('views/cannot_transfer_dialog');
+  var DevicePickerPanel = require('views/device_picker_panel');
+  var TurnBluetoothOnDialog = require('views/turn_bluetooth_on_dialog');
+
+  var _debug = false;
+  var Debug = function() {};
+  if (_debug) {
+    Debug = function bttm_debug(msg) {
+      console.log('--> [TransferManager]: ' + msg);
+    };
+  }
+
+  /*
+   * TransferManager is responsible for:
+   *   1. Handling system message 'activity' while there is an activity request
+   *      incoming. Then, it will post result to requesting app.
+   */
+  var TransferManager = {
+    /**
+     * An instance to access activity source.
+     *
+     * @access private
+     * @memberOf TransferManager
+     * @type {Object}
+     */
+    _activity: null,
+
+    /**
+     * An number to maintain remaining tasks in queue.
+     *
+     * @access private
+     * @memberOf TransferManager
+     * @type {Number}
+     */
+    _numberOfTasks: 0,
+
+    /**
+     * The init function only watches the 'activity' event from system message.
+     *
+     * @access private
+     * @memberOf TransferManager
+     */
+    init: function bttm_init() {
+      // Watch 'activity' event from system message.
+      this._watchActivityRequest();
+    },
+
+    /**
+     * Watch 'activity' event with callback function from system message.
+     *
+     * @access private
+     * @memberOf TransferManager
+     */
+    _watchActivityRequest: function bttm__watchActivityRequest() {
+      navigator.mozSetMessageHandler('activity',
+        this._activityHandler.bind(this));
+    },
+
+    /**
+     * Handle 'activity' request. Then make decision to service the request or
+     * reject it.
+     *
+     * @access private
+     * @memberOf TransferManager
+     * @param {Object} activityRequest
+     */
+    _activityHandler: function bttm__activityHandler(activityRequest) {
+      Debug('_activityHandler(): activityRequest = ' +
+            JSON.stringify(activityRequest));
+      if ((activityRequest.source.name == 'share' ||
```

Comment:
Not sure whether the checks on the activity names are required as we already listed them in the manifest.